File: src/main/java/org/displee/cache/index/archive/Archive317.java
package org.displee.cache.index.archive;

import org.displee.cache.index.archive.file.File;
import org.displee.io.impl.InputStream;
import org.displee.io.impl.OutputStream;
import org.displee.utilities.BZIP2Compressor;
import org.displee.utilities.Compression;
import org.displee.utilities.GZIPCompressor;

public class Archive317 extends Archive {

	private boolean extracted;

	private Compression.CompressionType compressionType;

	private int priority;

	public Archive317(int id) {
		super(id);
	}

	public Archive317(int id, int name) {
		super(id, name);
	}

	@Override
	public boolean read(InputStream s) {
		if (compressionType == Compression.CompressionType.GZIP) {
			setFileIds(1);
			setFiles(1);
			getFiles()[0] = new File(0, GZIPCompressor.deflate317(s.getBytes()));
			return read = true;
		}
		byte[] decompressed;
		int decompressedLength = s.read24BitInt();
		int compressedLength = s.read24BitInt();
		if (decompressedLength != compressedLength) {
			decompressed = BZIP2Compressor.decompress317(decompressedLength, compressedLength, s);
			extracted = true;
		} else {
			decompressed = s.readBytes(s.getRemaining());
		}

		InputStream stream = new InputStream(decompressed);
		int filesLength = stream.readUnsignedShort();
		setFileIds(filesLength);
		setFiles(filesLength);
		InputStream filesInputStream = new InputStream(stream.getBytes());
		filesInputStream.setOffset(stream.getOffset() + filesLength * 10);
		for(int i = 0; i < filesLength; i++) {
			int fileName = stream.readInt();
			decompressedLength = stream.read24BitInt();
			compressedLength = stream.read24BitInt();
			byte[] data;
			if (extracted) {
				data = filesInputStream.readBytes(decompressedLength);
			} else {
				data = BZIP2Compressor.decompress317(decompressedLength, compressedLength, filesInputStream);
			}
			getFileIds()[i] = i;
			getFiles()[i] = new File(i, data, fileName);
		}
		return read = true;
	}

	@Override
	public byte[] write(OutputStream outputStream) {
		if (compressionType == Compression.CompressionType.GZIP) {
			return GZIPCompressor.inflate317(getFiles()[0] == null ? new byte[0] : getFiles()[0].getData());
		}
		int length = getFiles().length;
		outputStream.writeShort(length);
		final OutputStream out = new OutputStream(2048);
		for(int i = 0; i < length; i++) {
			File file = getFiles()[i];
			outputStream.writeInt(file.getName());
			outputStream.write24BitInt(file.getData().length);
			byte[] toWrite = extracted ? file.getData() : BZIP2Compressor.compress(file.getData());
			outputStream.write24BitInt(toWrite.length);

			out.writeBytes(toWrite);
		}
		outputStream.writeBytes(out.flip());

		byte[] data = outputStream.flip();

		byte[] compressed = extracted ? BZIP2Compressor.compress(data) : data;

		final OutputStream finalStream = new OutputStream(compressed.length + 6);
		finalStream.write24BitInt(data.length);
		finalStream.write24BitInt(compressed.length);

		finalStream.writeBytes(compressed);
		return finalStream.flip();
	}

	public Compression.CompressionType getCompressionType() {
		return compressionType;
	}

	public void setCompressionType(Compression.CompressionType compressionType) {
		this.compressionType = compressionType;
	}

	public int getPriority() {
		return priority;
	}

	public void setPriority(int priority) {
		this.priority = priority;
	}

}

File: src/main/java/org/displee/cache/index/archive/Archive317.kt
package org.displee.cache.index.archive

import org.displee.cache.index.archive.file.File
import org.displee.io.impl.InputStream
import org.displee.io.impl.OutputStream
import org.displee.utilities.BZIP2Compressor
import org.displee.utilities.Compression.CompressionType
import org.displee.utilities.GZIPCompressor

class Archive317 : Archive {

    private var extracted = false
    var compressionType: CompressionType? = null
    var priority = 0

    constructor(id: Int) : super(id)
    constructor(id: Int, name: Int) : super(id, name)

    override fun read(inputStream: InputStream): Boolean {
        read = true
        if (compressionType == CompressionType.GZIP) {
            first()?.data = GZIPCompressor.deflate317(inputStream.bytes)
            return true
        }
        val decompressed: ByteArray
        var decompressedLength = inputStream.read24BitInt()
        var compressedLength = inputStream.read24BitInt()
        if (decompressedLength != compressedLength) {
            decompressed = BZIP2Compressor.decompress317(decompressedLength, compressedLength, inputStream)
            extracted = true
        } else {
            decompressed = inputStream.readBytes(inputStream.remaining)
        }
        val stream = InputStream(decompressed)
        val filesLength = stream.readUnsignedShort()
        val filesInputStream = InputStream(stream.bytes)
        filesInputStream.offset = stream.offset + filesLength * 10
        for (i in 0 until filesLength) {
            val fileName = stream.readInt()
            decompressedLength = stream.read24BitInt()
            compressedLength = stream.read24BitInt()
            val data: ByteArray = if (extracted) {
                filesInputStream.readBytes(decompressedLength)
            } else {
                BZIP2Compressor.decompress317(decompressedLength, compressedLength, filesInputStream)
            }
            files[i] = File(i, data, fileName)
        }
        return true
    }

    override fun write(): ByteArray {
        if (compressionType == CompressionType.GZIP) {
            val first = first()
            return GZIPCompressor.inflate317(if (first == null) ByteArray(0) else first.data)
        }
        val outputStream = OutputStream(10 * files.size)
        val files: Collection<File> = files.values
        outputStream.writeShort(files.size)
        val out = OutputStream(2048)
        for (file in files) {
            val fileData = file.data ?: continue
            outputStream.writeInt(file.hashName)
            outputStream.write24BitInt(fileData.size)
            val toWrite = if (extracted) fileData else BZIP2Compressor.compress(file.data)
            outputStream.write24BitInt(toWrite.size)
            out.writeBytes(toWrite)
        }
        outputStream.writeBytes(out.flip())
        val data = outputStream.flip()
        val compressed = if (extracted) BZIP2Compressor.compress(data) else data
        val finalStream = OutputStream(compressed.size + 6)
        finalStream.write24BitInt(data.size)
        finalStream.write24BitInt(compressed.size)
        finalStream.writeBytes(compressed)
        return finalStream.flip()
    }

}